sdk/core/azure-core/src/uuid.cpp

```diff
@@ -18,84 +23,158 @@ static thread_local std::mt19937_64 randomGenerator(std::random_device{}());
 } // namespace
 #endif
 
+using Azure::Core::_internal::StringExtensions;
+
 namespace {
-static char ByteToHexChar(uint8_t byte)
+/*
+"00000000-0000-0000-0000-000000000000"
+         ^    ^    ^    ^
+ 000000000011111111112222222222333333
+ 012345678901234567890123456789012345
+ \______________ = 36 ______________/
+*/
+constexpr size_t UuidStringLength = 36;
+constexpr bool IsDashIndex(size_t i) { return i == 8 || i == 13 || i == 18 || i == 23; }
+
+constexpr std::uint8_t HexToNibble(char c) // does not check for errors
 {
-  if (byte <= 9)
+  if (c >= 'a')
   {
-    return '0' + byte;
+    return 10 + (c - 'a');
   }
 
-  AZURE_ASSERT_MSG(
-      byte >= 10 && byte <= 15,
-      "It is expected, for a valid Uuid, to have byte values, where each of the two nibbles fit "
-      "into a hexadecimal character");
+  if (c >= 'A')
+  {
+    return 10 + (c - 'A');
+  }
 
-  return 'a' + (byte - 10);
+  return c - '0';
+}
+
+constexpr char NibbleToHex(std::uint8_t nibble) // does not check for errors
```

Comment:
Was it necessary to remove the azure_assert to make it a constexpr?
If not, than we should keep that invariant check in code, rather than a comment.

Reply:
And do we need it to be a constexpr?

Reply:
Looking at your other comments:
> AZURE_ASSERT() seems like an overkill to me, because it does compile in release configurations as well,...

I see. That makes sense to not use in loops, to enforce a variant we know/expect to be true.

Reply:
Yes, that is for the loop reason. Everything at the moment is pretty scoped here, plus it is covered with unit tests, so I think making the function to check for errors in a loop would be a bit overkill, not to mention the fact that it will be there for release builds as well.

As for the constexpr - it does not have to be constexpr, but if it is constexpr, it is better - it allows the compiler to do more optimizations, potentially producing more effective code.